metrics/exporter_test.go

```diff
@@ -102,22 +103,150 @@ func TestMain(m *testing.M) {
 }
 
 func TestMetricsExporter(t *testing.T) {
-	_, err := newMetricsExporter(&metricsConfig{
-		domain:               servingDomain,
-		component:            testComponent,
-		backendDestination:   "unsupported",
-		stackdriverProjectID: ""}, TestLogger(t))
-	if err == nil {
-		t.Errorf("Expected an error for unsupported backend %v", err)
-	}
+	tests := []struct {
+		name          string
+		config        *metricsConfig
+		expectSuccess bool
+	}{{
+		name: "unsupportedBackend",
+		config: &metricsConfig{
+			domain:             servingDomain,
+			component:          testComponent,
+			backendDestination: "unsupported",
+			stackdriverClientConfig: stackdriverClientConfig{
+				ProjectID: "",
+			},
+		},
+		expectSuccess: false,
+	}, {
+		name: "validConfig",
+		config: &metricsConfig{
+			domain:             servingDomain,
+			component:          testComponent,
+			backendDestination: Stackdriver,
+			stackdriverClientConfig: stackdriverClientConfig{
+				ProjectID: "testProj",
+			},
+		},
+		expectSuccess: true,
+	}, {
+		name: "stackdriverConfigOnly",
+		config: &metricsConfig{
+			backendDestination: Stackdriver,
+			stackdriverClientConfig: stackdriverClientConfig{
+				ProjectID:          "project",
+				GCPLocation:        "us-west1",
+				ClusterName:        "cluster",
+				GCPSecretName:      "secret",
+				GCPSecretNamespace: "secret-ns",
+			},
+		},
+		expectSuccess: true,
+	}, {
+		name: "fullValidConfig",
+		config: &metricsConfig{
+			domain:                            servingDomain,
+			component:                         testComponent,
+			backendDestination:                Stackdriver,
+			reportingPeriod:                   60 * time.Second,
+			isStackdriverBackend:              true,
+			stackdriverMetricTypePrefix:       path.Join(servingDomain, testComponent),
+			stackdriverCustomMetricTypePrefix: path.Join(customMetricTypePrefix, defaultCustomMetricSubDomain, testComponent),
+			stackdriverCustomMetricsSubDomain: defaultCustomMetricSubDomain,
+			stackdriverClientConfig: stackdriverClientConfig{
+				ProjectID:          "project",
+				GCPLocation:        "us-west1",
+				ClusterName:        "cluster",
+				GCPSecretName:      "secret",
+				GCPSecretNamespace: "secret-ns",
+			},
+		},
+		expectSuccess: true,
+	}, {
+		name: "prometheusBackendWithStackdriverConfig",
+		config: &metricsConfig{
+			domain:             servingDomain,
+			component:          testComponent,
+			backendDestination: Prometheus,
+			reportingPeriod:    5 * time.Second,
+			prometheusPort:     defaultPrometheusPort,
+			stackdriverClientConfig: stackdriverClientConfig{
+				ProjectID:          "project",
+				GCPLocation:        "us-west1",
+				ClusterName:        "cluster",
+				GCPSecretName:      "secret",
+				GCPSecretNamespace: "secret-ns",
+			},
+		},
+		expectSuccess: true,
+	}, {
+		name: "invalidStackdriverGcpLocation",
```

Comment:
Why is the location invalid? What does this test do?

Reply:
This is to make sure that creating an exporter does not fail if an invalid GCP location is used. GCP locations are restricted to a specified list and Stackdriver will reject requests with an invalid GCP location. 

However, opencensus does not validate and allows the exporter to be created. This behavior is essentially verified by this test. (Stackdriver will send back a descriptive error when the first metric is exported).

Reply:
Could you add comment to explain it?